### PR TITLE
Refactor then/success/failure

### PR DESCRIPTION
**Pull request by andr3ej**
to make ListenableFuture replacement less fiddly.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/Utils.kt [hunk 1]
```diff
@@ -74,38 +73,20 @@ fun <T> Future<T>.getOrThrow(timeout: Duration? = null): T {
...
-        return@then
-    }
-    body(r)
-}
+fun <F : ListenableFuture<*>, V> F.then(block: (F) -> V) = addListener(Runnable { block(this) }, MoreExecutors.directExecutor())
 
-fun <T> ListenableFuture<T>.failure(executor: Executor, body: (Throwable) -> Unit) = then(executor) {
-    try {
+fun <U, V> Future<U>.match(success: (U) -> V, failure: (Throwable) -> V): V {
+    return success(try {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: This is kind of an odd way to write it. What was wrong with:

```
try {
   return success(getOrThrow())
} catch (t: Throwable) {
   return failure(t)
}
```

Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: That would run failure if success breaks, but the expectation is getOrThrow successful => run success and getOrThrow fails => run failure.

**Comment by andr3ej** [comment]
Mostly separation of concerns and developer guidance:
* success and failure had an implicit "then", but often you know the future is done so there is no need to go via the listener API. The "then" is too subtle - I didn't notice it in ConcurrencyUtils. If you used both success and failure that's 2 listeners, with nondeterminism implications. NodeTerminalView used all 3 of then/success/failure, whereas now what actually happens under the hood is in line with the methods you would call if you were writing then out manually.
* The old success/failure didn't give you the option of blocking.
* match calls get() once, success+failure called it twice, which is potentially 2 ExecutionExceptions with their expensive stacktraces.
* The JDK won't do anything about the failure in the Future. If you used success on its own, you were wilfully neither logging nor propagating that failure, while the call site made it look like all was well. With match it's explicit that you're dropping the failure.
* match is consistent with our ErrorOr API.
* In StabilityTest and CrossCashTest did the author really mean to print out the future object in the success case? I think match makes it slightly harder to introduce that potential bug.
* then and match compose nicely into thenMatch.

File: core/src/main/kotlin/net/corda/core/Utils.kt (hunk 1 above)
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/Utils.kt]: Good point.

**Pull request merged**